### PR TITLE
Clean theme data when switching themes in the customizer

### LINKED ISSUE #34531: theme.json styles are not loaded when previewing a FSE theme in the Customizer
### Description

If you have a non-FSE theme active, go to the Customizer, click the "Change" button next to the name of the active theme, and select a FSE theme to preview, the preview that gets loaded doesn't include any of the theme.json styles in the theme.

The preview does work if you have a FSE theme active (like Blockbase) and preview a different FSE theme (like TT1 Blocks), but not if your active theme is a non-FSE theme.

### Step-by-step reproduction instructions

1. Activate a non-FSE theme, like Twenty Twenty-One.
2. Go to the Customizer.
3. Next to the theme name, click the "Change" button and select a FSE theme, like TT1 Blocks.
4. The preview is loaded, but the theme.json styles are not output, resulting in the theme looking broken.

### Screenshots, screen recording, code snippet

image

### Environment info

WordPress 5.8, Gutenberg 11.4.0

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by oandregal**
Fixes https://github.com/WordPress/gutenberg/issues/34531
Related https://github.com/WordPress/gutenberg/pull/34704
Related https://meta.trac.wordpress.org/ticket/5818#comment:4

When we read the `theme.json` from the theme, we cache its data and only clean it upon certain actions, such as `switch_theme`. In the customizer, the user can dynamically switch themes, which appears not to use the `switch_theme`. We still have the `setup_theme` and specific ones such as `start_previewing_theme`.

### CODE REVIEW

File: lib/class-wp-theme-json-resolver-gutenberg.php [hunk 1]
```diff
@@ -531,3 +531,4 @@ public static function clean_cached_data() {
 }
 
 add_action( 'switch_theme', array( 'WP_Theme_JSON_Resolver_Gutenberg', 'clean_cached_data' ) );
+add_action( 'setup_theme', array( 'WP_Theme_JSON_Resolver_Gutenberg', 'clean_cached_data' ) );
```
Comment (oandregal) [inline lib/class-wp-theme-json-resolver-gutenberg.php]: I wonder if it'd be enough to only hook to `setup_theme`. In other words: is there any situation in which the `switch_theme` is fired but the `setup_theme` isn't?

The current fix works, though, and the worst-case scenario is that we read the data from the sources twice for a theme.

Comment (ntsekouras) [reply lib/class-wp-theme-json-resolver-gutenberg.php]: I believe `setup_theme` alone could be enough. [switch_theme](https://developer.wordpress.org/reference/hooks/switch_theme/) is triggered when we deactivate a theme, which is not the case for the previews.

Maybe @aristath, @carolinan  or @swissspidy could confirm? 😄

Comment (swissspidy) [reply lib/class-wp-theme-json-resolver-gutenberg.php]: `setup_theme` is fired before the current theme is loaded when accessing WordPress (happens late in `wp-settings.php`). This only happens once per request. This is similar to the `plugins_loaded` hook.

`switch_theme()` (which fires the `switch_theme` action) can be called later on and potentially multiple times during a request, for example when publishing customizer changesets, upgrading the theme, or of course when switching themes.

For this reason, it's definitely needed to hook into `switch_theme`

Not super familiar with the Customizer, but `WP_Customize_Manager` hooks into `setup_theme` to preview different themes when accessing WordPress.

That makes sense, since the Customizer loads the WP site in an iframe and reloads that page when changing themes. So each time it's a new request, firing `setup_theme` anew, then kicking off the customizer logic.

Hooking into `setup_theme` to clean cached data seems a bit odd to do this.

In a regular request, there's no data to clear yet, and for customizer requests it just so happens that your code runs before the customizer.

It would be better to hook into the customizer logic itself.

**tl;dr:**

1. Hook into `switch_theme`
2. Hook into `start_previewing_theme`

Comment (oandregal) [reply lib/class-wp-theme-json-resolver-gutenberg.php]: Thanks for the explanation, Pascal! I've implemented what you suggested.

I agree using `setup_theme` is a bit odd (cleaning the cache before the theme data is even used, in most cases). However, I wonder if that's the hook that's most relevant to what we want to do here, instead of hooking into flow-specific ones. There may be flows that are not covered yet. I'll keep an eye on this.

Comment (oandregal) [reply lib/class-wp-theme-json-resolver-gutenberg.php]: I've run into a meta trac ticket that may be related to this https://meta.trac.wordpress.org/ticket/5818#comment:4 (previews for themes with theme.json not working consistently in the theme directory).

**Comment by oandregal** [comment]
Related PR at https://github.com/WordPress/gutenberg/pull/34704

**Review by ntsekouras — APPROVED** [review]
This works well! Thanks @oandregal and @swissspidy 💯

**Review by carolinan — APPROVED** [review]
I was able to reproduce the problem and the PR solves it for me.

**Review by aristath — APPROVED** [review]
